### PR TITLE
feat(schemas): Model most of input validation as schema

### PR DESCRIPTION
**Pull request by alex-hofsteede**
This takes most of the imperative validation code in validate_data
and uses declarative JSON schema to validate it instead. There is an
associated helper method that uses the schema validation errors to do
the same things as the old code, namely deleting invalid keys, or
defaulting them if there is an applicable default.

There is a pre-validation casting step that forces some of the values to be
certain types before we apply the schema validation. This is to avoid
having to create a bunch of special cases in the schema for eg. values
that are either integers OR numeric strings. In theory this casting code
could be removed completely if all clients were well behaved and only
sent data for a given key in a single correct format.

There are also additional validators for the individual interfaces that
are created. These are checked when the interface is created with
to_python.

I had to make minor changes to some tests as the choice to raise an
APIError on some types of invalid keys (eg. non-string event_id value)
but not on others (culprit too long, too short, etc) seemed kinda
arbitrary anyway and it would have made for some messy special case code
to fix.

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by alex-hofsteede** [comment]
Requesting final review on this. Have compared against a ton of live traffic. The current run of the comparer shows only 19 records different after running `validate_data`, out of 3408182 compared so far, which is `0.0006%`. The differences are all expected given the schema.

**Review by mitsuhiko — APPROVED** [review]
ship it!

File: src/sentry/conf/server.py [hunk 1]
```diff
@@ -802,6 +802,7 @@ def create_partitioned_queues(name):
     'request': 'sentry.interfaces.http.Http',
     'sdk': 'sentry.interfaces.sdk.Sdk',
     'stacktrace': 'sentry.interfaces.stacktrace.Stacktrace',
+    'frame': 'sentry.interfaces.stacktrace.Frame',
```
Comment (dcramer) [inline src/sentry/conf/server.py]: if we dont special case this (i didnt see) then it means someone could send this attribute at the top level of the payload

Comment (mitsuhiko) [reply src/sentry/conf/server.py]: Isn't that already the case?

Comment (mitsuhiko) [reply src/sentry/conf/server.py]: Oh sorry, missed that. Yeah, that should not go there.

Comment (alex-hofsteede) [reply src/sentry/conf/server.py]: gotcha, will fix

**Pull request merged**